db.py

```diff
@@ -284,10 +305,68 @@ def send_links(*, email, planttracer_endpoint):
                             msg=msg
                             )
 
+################ API KEY ################
+
+def make_new_api_key(*,email):
+    """Create a new api_key for an email that is registered
+    :param: email - the email
+    :return: api_key - the api_key
+    """
+    user = lookup_user(email=email)
+    if user and user['enabled'] == 1:
+        user_id = user['id']
+        api_key = str(uuid.uuid4()).replace('-', '')
+        dbfile.DBMySQL.csfr(get_dbwriter(),
+                                """INSERT INTO api_keys (user_id, api_key) VALUES (%s,%s)""",
+
+                                (user_id, api_key))
+        # Manually log so that the api_key is not logged
+        logit(func_name='make_new_api_key', func_args={'email': email}, func_return={'api_key':'*****'})
+        return api_key
+    return None
+
+
+@log
+def delete_api_key(api_key):
+    """Deletes an api_key
+    :param: api_key - the api_key
+    :return: the number of keys deleted
+    """
+    if len(api_key) < 10:
+        raise InvalidAPI_Key(api_key)
+    return dbfile.DBMySQL.csfr(get_dbwriter(),
+                               """DELETE FROM api_keys WHERE api_key=%s""",
+                               (api_key,))
 
-################################################################
-# Course Management
-################################################################
+
+@log
+def list_users(*, user_id):
+    """Returns a directory of all the courses to which the user has access, and all of the people in them.
+    :param: user_id - the user doing the listing (determines what they can see)
+    """
+    cmd = """SELECT users.name AS name,users.email AS email,users.primary_course_id as primary_course_id, users.id AS user_id,
+                    k.first as first,k.last as last
+              FROM users LEFT JOIN
+                      (select user_id,min(first_used_at) as first,max(last_used_at) as last from api_keys) k
+              ON users.id=k.user_id
+              WHERE users.id=%s
+                OR users.primary_course_id IN (select primary_course_id from users where id=%s)
+                OR users.primary_course_id IN (select course_id from admins where user_id=%s)
+                OR %s IN (select user_id from admins where course_id=%s)
+              ORDER BY primary_course_id,name,email"""
+    args = (user_id, user_id,user_id,user_id,SUPER_ADMIN_COURSE_ID)
+    return dbfile.DBMySQL.csfr(get_dbreader(),cmd,args,asDicts=True)
+
```

Comment:
This select query is returning no results for me when I set %s to 2 (sbarber2@gmail.com) or 5 (sbarber2+admin@gmail.com). If I take out the LEFT JOIN the select does return all 19 rows, that is, all the users in course_id 1. So I think something is broken here.

Reply:
select user_id,min(first_used_at) as first,max(last_used_at) as last from api_keys where user_id = 2;
also returns something reasonable looking. So I think something is going on with that LEFT JOIN to empty out the result set of the top-level SELECT, but I haven't figured out what.

Reply:
Oh, I see the what might fix this: add a GROUP BY user_id to the table subquery on line 350:

select user_id,min(first_used_at) as first,max(last_used_at) as last from api_keys group by user_id

Then:

SELECT users.name AS name,users.email AS email,users.primary_course_id as primary_course_id, users.id AS user_id,
                    k.first as first,k.last as last
              FROM users LEFT JOIN
                      (select user_id,min(first_used_at) as first,max(last_used_at) as last from api_keys group by user_id) k
              ON users.id=k.user_id
              WHERE users.id=2
                OR users.primary_course_id IN (select primary_course_id from users where id=2)
                OR users.primary_course_id IN (select course_id from admins where user_id=2)
                OR 2 IN (select user_id from admins where course_id=2)
              ORDER BY primary_course_id,name,email

Gives the expected 19 row result.